src/impl_methods.rs

```diff
@@ -2605,15 +2617,14 @@ where
                 unsafe { unlimited_transmute::<B, A>(b) }
             };
             // Delegate to mapv_into() using the wrapped closure.
-            // Convert output to a uniquely owned array of type Array<A, D>.
-            let output = self.mapv_into(f).into_owned();
-            // Change the return type from Array<A, D> to Array<B, D>.
+            let output = self.mapv_into(f);
+            // Change the return type of the data from A to B.
             // Again, safe because A and B are the same type.
-            unsafe { unlimited_transmute::<Array<A, D>, Array<B, D>>(output) }
+            unsafe { unlimited_transmute::<ArrayBase<S, D>, ArrayBase<T, D>>(output) }
```

Comment:
It looks like S and T can be completely different types, can we show how this would be safe?

Reply:
Oops, you are right! 😳 The array representation of `S` might not be the same as the array representation of `T`, so even thought the element types `A` and `B` are the same, the types `S` and `T` might not be the same.

Fortunately, there is no type-related problem in Rust that cannot be solved with more traits!  Using the `Plug` trait from [here](https://github.com/edmundsmith/type-plugs) we can introspect on the type of the data representation (e.g. `OwnedRepr` vs `OwnedArcRepr`) and perform the conversion soundly.  I would appreciate feedback on where to put the `Plug` trait and whether it is acceptable to expose such a trait through the API at all.

Reply:
It's not on crates.io so it cannot be a dependency. I will decline to look into this unfortunately

Reply:
I think the idea is add a plug trait as proposed here in the current revision already, not add a dependency on some crate. The linked repository has an explanatory character as well which is why it isn't a published crate if I understand this correctly.

So there is very little code involved to establish the relevant mappings on a type level, but the resulting API is definitely on the complex end of the scale w.r.t. its type bounds and hence probably not particularly simple to explain.